packages/backend/src/api/controllers/TransactionSubmitController.ts

```diff
@@ -144,50 +141,172 @@ export class TransactionSubmitController {
+    if (!fetched.isSuccess) {
+      return fetched.controllerResult
+    }
+    if (!tradeMatchesOffer(offer, fetched.data)) {
+      return { type: 'bad request', message: `Trade does not match offer` }
     }
+
     // TODO: cross repository transaction
     await this.offersRepository.updateTransactionHash(offerId, transactionHash)
     await this.sentTransactionRepository.add({
       transactionHash,
       timestamp,
       data: {
         type: 'ForcedTrade',
-        starkKeyA: data.starkKeyA,
-        starkKeyB: data.starkKeyB,
-        positionIdA: data.positionIdA,
-        positionIdB: data.positionIdB,
-        collateralAmount: data.collateralAmount,
+        starkKeyA: fetched.data.starkKeyA,
+        starkKeyB: fetched.data.starkKeyB,
+        positionIdA: fetched.data.positionIdA,
+        positionIdB: fetched.data.positionIdB,
+        collateralAmount: fetched.data.collateralAmount,
         collateralAssetId: this.collateralAsset.assetId,
-        syntheticAmount: data.syntheticAmount,
-        syntheticAssetId: data.syntheticAssetId,
-        isABuyingSynthetic: data.isABuyingSynthetic,
-        submissionExpirationTime: data.submissionExpirationTime,
-        nonce: data.nonce,
-        signatureB: data.signature,
-        premiumCost: data.premiumCost,
+        syntheticAmount: fetched.data.syntheticAmount,
+        syntheticAssetId: fetched.data.syntheticAssetId,
+        isABuyingSynthetic: fetched.data.isABuyingSynthetic,
+        submissionExpirationTime: fetched.data.submissionExpirationTime,
+        nonce: fetched.data.nonce,
+        signatureB: fetched.data.signature,
+        premiumCost: fetched.data.premiumCost,
         offerId,
       },
     })
     return { type: 'created', content: { id: transactionHash } }
   }
+
+  async submitFreezeRequest(
+    transactionHash: Hash256
+  ): Promise<ControllerResult> {
+    const timestamp = Timestamp.now()
+    const fetched = await this.fetchTxAndDecode(
+      transactionHash,
+      this.contracts.perpetual,
+      decodeFreezeRequest
+    )
+    if (!fetched.isSuccess) {
+      return fetched.controllerResult
+    }
+
+    await this.sentTransactionRepository.add({
+      transactionHash,
+      timestamp,
+      data: {
+        type: 'FreezeRequest',
+        starkKey: fetched.data.starkKey,
+        positionOrVaultId: fetched.data.positionOrVaultId,
+        quantizedAmount: fetched.data.quantizedAmount,
+      },
+    })
+    return { type: 'created', content: { id: transactionHash } }
+  }
+
+  async submitEscapeVerified(
+    transactionHash: Hash256,
+    starkKey: StarkKey,
+    positionOrVaultId: bigint
+  ): Promise<ControllerResult> {
+    const timestamp = Timestamp.now()
+    const fetched = await this.fetchTxAndDecode(
+      transactionHash,
+      this.contracts.escapeVerifier,
+      validateVerifyEscapeRequest
+    )
+    if (!fetched.isSuccess) {
+      return fetched.controllerResult
+    }
+
```

Comment:
You still need to check if fetched.data is true or false

Reply:
No need, in verifyAndDecodeTx fn there is a check for this: !data